src/test/app/Batch_test.cpp

```diff
@@ -0,0 +1,3047 @@
+        }
+
+        // tec failure
+        {
+            auto const preAlice = env.balance(alice);
+            auto const preBob = env.balance(bob);
+
+            auto const batchFee = batch::calcBatchFee(env, 0, 2);
+            auto const seq = env.seq(alice);
+            env(batch::outer(alice, seq, batchFee, tfAllOrNothing),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 1),
+                batch::inner(pay(alice, bob, XRP(999)), seq + 2),
+                ter(tesSUCCESS));
+            auto const txIDs = env.tx()->getBatchTransactionIDs();
+            TxID const parentBatchId = env.tx()->getTransactionID();
+            std::vector<TestBatchData> testCases = {};
+            env.close();
+            validateBatch(env, parentBatchId, testCases);
+
+            // Alice consumes sequence
+            BEAST_EXPECT(env.seq(alice) == seq + 1);
+
+            // Alice pays Fee; Bob should not be affected
+            BEAST_EXPECT(env.balance(alice) == preAlice - batchFee);
+            BEAST_EXPECT(env.balance(bob) == preBob);
+        }
+
+        // non tec failure
+        {
+            auto const preAlice = env.balance(alice);
+            auto const preBob = env.balance(bob);
+
+            auto const batchFee = batch::calcBatchFee(env, 0, 2);
+            auto const seq = env.seq(alice);
+            auto tx2 = trust(alice, USD(1000), tfSetfAuth);
+            tx2[jss::Fee] = "0";
+            tx2[jss::Sequence] = seq + 2;
+            tx2[jss::SigningPubKey] = "";
+            tx2[jss::Flags] = tx2[jss::Flags].asUInt() | tfInnerBatchTxn;
+            env(batch::outer(alice, seq, batchFee, tfAllOrNothing),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 1),
+                batch::inner_nofill(tx2),
+                ter(tesSUCCESS));
+            auto const txIDs = env.tx()->getBatchTransactionIDs();
+            TxID const parentBatchId = env.tx()->getTransactionID();
+            std::vector<TestBatchData> testCases = {
+                // tx #2 fails with tefNO_AUTH_REQUIRED
+            };
+            env.close();
+            validateBatch(env, parentBatchId, testCases);
+
+            // Alice consumes sequence
+            BEAST_EXPECT(env.seq(alice) == seq + 1);
+
+            // Alice pays Fee; Bob should not be affected
+            BEAST_EXPECT(env.balance(alice) == preAlice - batchFee);
+            BEAST_EXPECT(env.balance(bob) == preBob);
+        }
+    }
+
+    void
+    testOnlyOne(FeatureBitset features)
+    {
+        testcase("only one");
+
+        using namespace test::jtx;
+        using namespace std::literals;
+
+        test::jtx::Env env{*this, envconfig()};
+
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+        auto const gw = Account("gw");
+        auto const USD = gw["USD"];
+        env.fund(XRP(1000), alice, bob, gw);
+        env.close();
+
+        // tec failure
+        {
```

Comment:
Please add:
* A test where the first transaction succeeds
* A test where all transactions fail